xarray_beam/_src/pangeo_test.py

```diff
@@ -0,0 +1,219 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for xarray_beam._src.pangeo."""
+
+import contextlib
+import itertools
+import tempfile
+
+import numpy as np
+from pangeo_forge_recipes.patterns import FilePattern, ConcatDim
+
+from xarray_beam._src import core
+from xarray_beam._src import test_util
+from xarray_beam._src.pangeo import FilePatternToChunks, _zero_dimensions, \
+  _expand_dimensions_by_key
+
+
+class ExpandDimensionsByKeyTest(test_util.TestCase):
+  TEST_DATA = test_util.dummy_era5_surface_dataset()
+  LEVEL = ConcatDim("level", list(range(91, 100)))
+  PATTERN = FilePattern(lambda level: f"gs://dir/{level}.nc", LEVEL)
+
+  def test_expands_dimensions(self):
+    key = core.Key(offsets={"time": 0, "level": 0})
+
+    for i, (index, _) in enumerate(self.PATTERN.items()):
+      actual = _expand_dimensions_by_key(
+        self.TEST_DATA, key, index, self.PATTERN
+      )
+
+      expected_dims = dict(self.TEST_DATA.dims)
+      expected_dims.update({"level": 1})
+
+      self.assertEqual(expected_dims, dict(actual.dims))
+      self.assertEqual(np.array([self.LEVEL.keys[i]]), actual["level"])
+
+  def test_raises_error_when_dataset_is_not_found(self):
+    key = core.Key({"time": 0, "boat": 0})
+    index = (0,)
+    with self.assertRaises(ValueError) as e:
+      _expand_dimensions_by_key(
+        self.TEST_DATA, key, index, self.PATTERN
+      )
+    self.assertIn("boat", e.exception.args[0])
+
+
+class FilePatternToChunksTest(test_util.TestCase):
+  TEST_DATA = test_util.dummy_era5_surface_dataset()
+
+  @contextlib.contextmanager
+  def pattern_from_testdata(self, test_data=None) -> FilePattern:
+    """Produces a FilePattern for a temporary NetCDF file of test data."""
+    if test_data is None:
+      test_data = self.TEST_DATA
+
+    try:
+      with tempfile.TemporaryDirectory() as tmpdir:
+        target = f'{tmpdir}/era5.nc'
+        test_data.to_netcdf(target)
+        yield FilePattern(lambda: target)
+    finally:
+      pass
```

Comment:
This `finally` clause looks like it doesn't do anything? If that's right, remove it?

Reply:
When I remove it, I get an error that says "expected except or finally block". I think I need to keep it.

Reply:
You would need to remove the `try:` part, too.